### PR TITLE
Cloudflare refactory with Terraform

### LINKED ISSUE #35: Cloudflare management should be moved to Terraform
Cloudflare management should be moved to Terraform. It is easy to forget to remove DNS records after `terraform destroy`.

### PR DESCRIPTION
**Pull request by andersla**
## Change content and motivation
Setting Cloudflare domain with Terraform module instead of Ansible. It is very easy to forget to run the playbook when destroying.

## GitHub cross-links 

**Fixes:** fixes #35 

**Docs**:  kubenow/docs#12

### CODE REVIEW

File: aws/main.tf [hunk 1]
```diff
@@ -157,14 +163,31 @@ module "edge" {
   master_ip = "${element(module.master.local_ip_v4, 0)}"
 }
 
+#
```
Comment (mcapuccini) [inline aws/main.tf]: Please use single # for comments: `# The code below (from here to end) should be identical for all cloud providers`

File: aws/main.tf [hunk 2]
```diff
@@ -157,14 +163,31 @@ module "edge" {
...
+# The code below (from here to end) should be identical for all cloud providers
+#
+
+# set cloudflare record (optional)
+module "cloudflare" {
+  # count values can not be dynamically computed, that's why using
```
Comment (mcapuccini) [inline aws/main.tf]: that's why we are* using, also this can go in single line since the next lines are longer anyway

File: aws/main.tf [hunk 3]
```diff
@@ -157,14 +163,31 @@ module "edge" {
...
+  # concat lists (record_count is limiting master_ip:s from being added to cloudflare if var.master_as_edge=false)
+  # terraform interpolation is limited and can not return list in conditionals
+  iplist = "${concat(module.edge.public_ip, module.master.public_ip)}"
+}
+
 # Generate Ansible inventory (identical for each cloud provider)
 resource "null_resource" "generate-inventory" {
 
-  # Changes to any node IP trigger inventory rewrite
+  # Trigger rewrite of inventory, uuid() generates a random string everytime it is called
```
Comment (mcapuccini) [inline aws/main.tf]: this is to rewrite always right?

File: aws/main.tf [hunk 4]
```diff
@@ -182,7 +205,11 @@ resource "null_resource" "generate-inventory" {
   }
   # output the lists formated
   provisioner "local-exec" {
-    command =  "echo \"${join("\n",formatlist("%s ansible_ssh_host=%s ansible_ssh_user=ubuntu", concat(module.master.hostnames, module.edge.hostnames), concat(module.master.public_ip, module.edge.public_ip)))}\" >> inventory"
+    command =  "echo \"${join("\n",formatlist("%s ansible_ssh_host=%s ansible_ssh_user=ubuntu", module.edge.hostnames, module.edge.public_ip))}\" >> inventory"
+  }
+  # only output if master is edge
```
Comment (mcapuccini) [inline aws/main.tf]: nope, one could in principle have master_as_edge and use edge nodes at the same time. Can we output eges IPs and master node IP together?

File: aws/node/main.tf [hunk 5]
```diff
@@ -90,7 +90,7 @@ output "public_ip" {
   value = ["${aws_instance.instance.*.public_ip}"]
 }
 output "hostnames" {
-  value = ["${aws_instance.instance.*.tags.Name}"]
+  value = ["${split(" ", replace(join(" ",formatlist("ip-%s", aws_instance.instance.*.private_ip)),".","-"))}"]
```
Comment (mcapuccini) [inline aws/node/main.tf]: this mode shouldn't belong to Cloudflare, but it's fine as long as you write it in the description of the PR.

**Comment by mcapuccini** [comment]
@andersla CI failed probably due to timeout in OpenStack... trying to restart...

File: aws/main.tf (hunk 4 above)
Comment (mcapuccini) [inline aws/main.tf]: Also in this way you skip the edge node testing in the CI

File: aws/main.tf (hunk 1 above)
Comment (andersla) [inline aws/main.tf]: fixed

File: aws/main.tf (hunk 2 above)
Comment (andersla) [inline aws/main.tf]: fixed

File: aws/main.tf (hunk 3 above)
Comment (andersla) [inline aws/main.tf]: OK

File: aws/main.tf (hunk 4 above)
Comment (andersla) [inline aws/main.tf]: The provisioner just below takes care of that:)

File: aws/node/main.tf (hunk 5 above)
Comment (andersla) [inline aws/node/main.tf]: ok

**Comment by mcapuccini** [comment]
Thanks @andersla! Pinging @jonalv and @StaffanArvidsson for test... that's a long wanted feature 🙂

**Pull request merged**